PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+        lineage_index (int)
+            Label of current lineage.
+        lineage_seed (dict)
+            Dict of seeds used to initiate pre-existing lineage definitions.
+
+    Returns:
+        seed_isolate (str)
+            Isolate to used to initiate next lineage.
+
+    """
+    # variable to return
+    seed_isolate = None
+    # first test if there is an existing seed
+    if lineage_index in lineage_seed.keys():
+        original_seed_isolate = lineage_seed[lineage_index]
+        # if seed now belongs to a different lineage, then lineage has been overwritten
+        # seed may be unassigned if neighbours have changed - but this does not overwrite the
+        # lineage in the same way
+        if lineage_clustering[original_seed_isolate] == null_cluster_value or lineage_clustering[original_seed_isolate] == lineage_index:
+            seed_isolate = original_seed_isolate
+    # else identify a new seed from the closest pair
+    else:
+        for index,(distance,(isolate1,isolate2)) in enumerate(zip(distances,row_labels)):
+            if lineage_clustering[isolate1] == null_cluster_value:
+                seed_isolate = isolate1
+                break
+            elif lineage_clustering[isolate2] == null_cluster_value:
+                seed_isolate = isolate2
+                break
+    # return identified seed isolate
+    return seed_isolate
+
+def get_lineage_clustering_information(seed_isolate, row_labels, distances):
+    """ Generates the ranked distances needed for cluster
+    definition.
+    
+    Args:
+        seed_isolate (str)
+            Isolate used to initiate lineage.
+        row_labels (list of tuples)
+            Pairs of isolates labelling each distance.
+        distances (numpy array)
+            Pairwise distances used for defining relationships.
+    
+    Returns:
+        lineage_info (dict)
+            Dict listing isolates by ranked distance from seed.
+    
+    """
+    # data structure
+    lineage_info = defaultdict(list)
+    rank = 0
+    last_dist = -1
+    # iterate through distances recording rank
+    for index,distance in enumerate(distances):
+        if seed_isolate in row_labels[index]:
+            if distance > last_dist:
+                rank = rank + 1
+                last_dist = distance
+            pair = row_labels[index][0] if row_labels[index][1] == seed_isolate else row_labels[index][1]
+            lineage_info[rank].append(pair)
+    # return information
+    return lineage_info
+
+def generate_nearest_neighbours(distances, row_labels, isolate_list, R):
+    # data structures
+    nn = {}
+    last_dist = {}
+    num_ranks = {}
+    for i in isolate_list:
+        nn[i] = {}
+        num_ranks[i] = 0
+    total_isolates = len(isolate_list)
+    num_distances = len(distances)
+    completed_isolates = 0
+    index = 0
+    # iterate through distances until all nearest neighbours identified
+    while completed_isolates < total_isolates and index < num_distances:
```

Comment:
Instead of using the long form matrix, making it square and sorting each row would probably be easier (and faster)